Refactor: introduce details level enum

xzfc opened the pull request:
Replace raw `usize` with an enum for telemetry details level. So now we can use "find references" feature of IDEs to find all the places where details level is significant.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

timvisee reviewed (commented):
I like this approach!

Though, we may better use `>=` rather than `>`. So that if we do `>= Level1` we show it at level 1 and higher. That makes more sense to me, because that way you could jump to all places in code where level 1 (or higher) is used as you note in the PR description.

xzfc commented:
Updated the PR, now it's ready for review.
The updated version introduces a boolean value `TelemetryDetail::histograms` and passing of `TelemetryDetail` to every place that potentially could call `OperationDurationsAggregator::get_statistics()`.
It is not used right now, but will be used after merging #3552.

> Though, we may better use `>=` rather than `>`.

Thanks for the suggestion, I've updated the code.

timvisee commented on lib/common/common/src/types.rs:
```diff
@@ -26,3 +26,35 @@ impl PartialOrd for ScoredPointOffset {
     }
 }
+
+#[derive(Copy, Clone, Debug)]
+pub struct TelemetryDetail {
+    pub level: DetailsLevel,
+    pub histograms: bool,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DetailsLevel {
+    Level0,
+    Level1,
+    Level2,
+}
+
+impl Default for TelemetryDetail {
+    fn default() -> Self {
+        TelemetryDetail {
+            level: DetailsLevel::Level0,
+            histograms: false,
+        }
+    }
+}
+
+impl From<usize> for DetailsLevel {
+    fn from(value: usize) -> Self {
+        match value {
+            0 => DetailsLevel::Level0,
+            1 => DetailsLevel::Level1,
+            _ => DetailsLevel::Level2,
+        }
+    }
+}
```
We can derive `Default` like this, without any custom implementation:

```rust
#[derive(Copy, Clone, Debug, Default)]
pub struct TelemetryDetail {
    pub level: DetailsLevel,
    pub histograms: bool,
}

#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
pub enum DetailsLevel {
    #[default]
    Level0,
    Level1,
    Level2,
}
```

What do you think? I don't have a strong preference.

xzfc replied:
I'd prefer not to have `Default` at all in the production code so it'll be always specified explicitly at the top-level.
However, it's convenient for unit/integration tests where concrete values are not important.

So, as a compromise I'd prefer to keep `Default` for `TelemetryDetail` (used in a test), but not for `DetailsLevel`.

timvisee replied:
Check! But the we probably want to add `#[cfg(test)]` to the `impl Default` we have in this PR.

xzfc replied:
Items under `#[cfg(test)]` couldn't be used in integration tests as they're separate crates. :( Only unit tests could use them.

timvisee reviewed (approved):
As far as I can see, the level thresholds for everything are still the same, which would make it a great change :+1:
